block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```
@@ -133,57 +135,109 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
+
+        for {
+          // This filter is required to enable adding blocks backward from higher height to lower
+          newlyBondedUnseen <- newlyBondedSet.toList.filterA(
+                                lm => latestMessagesIndex.contains(lm).not
+                              )
+          newlyBondedLMs = newlyBondedUnseen.map(v => (v, block.blockHash)).toSet
+          // As genesis has empty sender, in the very first block sender will be in newlyBondedUnseen
+          // (no self justification). In this case LM should be updated. It interferes with protection from
+          // regressive seq number for LM, so needs some additional logic to check.
+          sendersLmIsGenesis <- for {
+                                 sendersCurLM <- latestMessagesIndex.get(block.sender)
+                                 r <- sendersCurLM match {
+                                       case None => false.pure[F]
+                                       case Some(lm) =>
+                                         blockMetadataIndex.getUnsafe(lm).map(_.seqNum).map(_ == 0)
+                                     }
+                               } yield r
+
+          // Do not create LM for empty sender, which is a now valid sender for genesis
+          lms <- if (senderIsEmpty || (newlyBondedSet.contains(block.sender) && !sendersLmIsGenesis))
+                  logEmptySender.whenA(senderIsEmpty) >> newlyBondedLMs.pure[F]
+                // Update LM for sender
+                else {
+                  // LM should be updated only when block sequence number advances.
+                  val senderCurLMSeqNum = latestMessagesIndex
+                    .getUnsafe(block.sender)
+                    .flatMap(blockMetadataIndex.getUnsafe)
+                    .map(_.seqNum)
+                  val checkRegressiveSeqNum  = senderCurLMSeqNum.map(block.seqNum < _)
+                  val senderHasInvalidFormat = block.sender.size() != Validator.Length
```

This check can be done immediately and check can fail fast before any other checks.

---

The thing is that empty sender is allowed, so `senderHasInvalidFormat` has to be in clause when LM is updated